# 기술 통계량을 계산하는 함수 정의

In [1]:
test <- read.csv("Rdata/test.csv")
head(test)

A B C D E
1 2 4 4 2 2
2 1 2 2 2 2
3 2 3 4 3 3
4 3 5 5 3 3
5 3 2 4 4 4
6 4 3 3 4 2

In [2]:
tail(test)

A B C D E
397 2 3 5 2 4
398 4 3 3 2 4
399 2 4 4 2 3
400 2 2 5 2 4
401 2 3 4 2 4
402 2 2 4 2 4

In [3]:
str(test)

'data.frame':	402 obs. of  5 variables:
 $ A: int  2 1 2 3 3 4 3 4 4 4 ...
 $ B: int  4 2 3 5 2 3 4 2 4 2 ...
 $ C: int  4 2 4 5 4 3 4 4 5 4 ...
 $ D: int  2 2 3 3 4 4 2 4 3 2 ...
 $ E: int  2 2 3 3 4 2 2 4 4 3 ...


In [4]:
summary(test)

       A               B               C               D              E        
 Min.   :1.000   Min.   :1.000   Min.   :1.000   Min.   :1.00   Min.   :1.000  
 1st Qu.:2.000   1st Qu.:2.000   1st Qu.:3.000   1st Qu.:2.00   1st Qu.:3.000  
 Median :3.000   Median :3.000   Median :4.000   Median :2.00   Median :4.000  
 Mean   :2.734   Mean   :2.908   Mean   :3.622   Mean   :2.51   Mean   :3.386  
 3rd Qu.:3.000   3rd Qu.:4.000   3rd Qu.:4.000   3rd Qu.:3.00   3rd Qu.:4.000  
 Max.   :5.000   Max.   :5.000   Max.   :5.000   Max.   :4.00   Max.   :5.000  

In [5]:
# A컬럼의 최댓값, 최솟값, 빈도수(table//프리퀀시?) 구하기
max(test$A)
min(test$A)
table(test$A)

[1] 5

[1] 1


  1   2   3   4   5 
 30 133 156  80   3 

In [6]:
# 각 컬럼의 빈도수 구하기(다 for문) //내가 한거

for(i in 1:5){
     cat(i,"번째 컬럼의 빈도수 분석 결과")
    print(table(test[i]))
    cat("------------------------\n")
}



1 번째 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
 30 133 156  80   3 
------------------------
2 번째 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
 16 150  98 131   7 
------------------------
3 번째 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
  3  74  72 176  77 
------------------------
4 번째 컬럼의 빈도수 분석 결과
  1   2   3   4 
 30 178 153  41 
------------------------
5 번째 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
  8  81 107 160  46 
------------------------


In [7]:
# 각 컬럼의 빈도수 구하기(다 for문) //선생님

#test[ ,1]
#ncol(test)
#colnames(test)
for(i in 1:ncol(test)){
   cat(colnames(test)[i], "컬럼의 빈도수 분석 결과")
    print(table(test[i]))
    cat("------------------------\n")
}

A 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
 30 133 156  80   3 
------------------------
B 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
 16 150  98 131   7 
------------------------
C 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
  3  74  72 176  77 
------------------------
D 컬럼의 빈도수 분석 결과
  1   2   3   4 
 30 178 153  41 
------------------------
E 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
  8  81 107 160  46 
------------------------


In [8]:
# 각컬럼 빈도수의 최댓값 최솟값 구하기
col <- 0
for(i in 1:ncol(test)){
    col <- table(test[i])
    #print(col)
    cat(colnames(test)[i], "컬럼의 빈도수의 최댓값과 최솟값 \n","최댓값 :",max(col),"최솟값 :", min(col),"\n")
    cat("------------------------\n")

}



A 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 156 최솟값 : 3 
------------------------
B 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 150 최솟값 : 7 
------------------------
C 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 176 최솟값 : 3 
------------------------
D 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 178 최솟값 : 30 
------------------------
E 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 160 최솟값 : 8 
------------------------


In [9]:
# 각 컬럼의 빈도수 구하기와 각 컬럼의 빈도수의 최댓값,최솟값 구하는 함수 만들기
# for문의 범위가 같으니까 하나만 있으면 됨


tableMaxMin <- function(x){
        for(i in 1:ncol(x)){
        cat(colnames(x)[i], "컬럼의 빈도수 분석 결과")
        print(table(x[i]))
        cat("------------------------\n")
        col <- table(x[i])
        cat(colnames(x)[i], "컬럼의 빈도수의 최댓값과 최솟값 \n","최댓값 :",max(col),"최솟값 :", min(col),"\n")
        cat("========================\n")
    }
}

tableMaxMin(test)

A 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
 30 133 156  80   3 
------------------------
A 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 156 최솟값 : 3 
B 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
 16 150  98 131   7 
------------------------
B 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 150 최솟값 : 7 
C 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
  3  74  72 176  77 
------------------------
C 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 176 최솟값 : 3 
D 컬럼의 빈도수 분석 결과
  1   2   3   4 
 30 178 153  41 
------------------------
D 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 178 최솟값 : 30 
E 컬럼의 빈도수 분석 결과
  1   2   3   4   5 
  8  81 107 160  46 
------------------------
E 컬럼의 빈도수의 최댓값과 최솟값 
 최댓값 : 160 최솟값 : 8 


---
# 결측치 (missing value) 자료 평균 계산


In [49]:
data <- c(1,2,3,4,5,6,7,8,9,NA,11,12,NA,14,15,NA,17,18,19,NA)

In [15]:
# NA가 있으면 mean 이 작동을 안함.
mean(data)

[1] NA

In [20]:
# NA 갯수 구하기
#data 크기에 따라 사용하는 코드가 같아야 잘 만든 코드임.
sum(is.na(data))


#또는 똑같은것의 갯수를 구하는것  = 빈도수 구하기 : table
#사용하기 좋은!!
table(is.na(data))

[1] 4


FALSE  TRUE 
   16     4 

In [42]:
# NA를 제외하고 평균 구하기
# 결측치를 무조건 제거하면 총계량을 얻을 수 없으며, 데이터의 손실로 문제 발생
# NA값에 30%이상이면 그 데이터는 못씀..
# 수집 > 저징 > 처리 > 분석 : 지금이 처리과정인데 데이터 버리면 다시 > 수집 앞에 과정이 오래걸림.
# 수집 >> 처리 까지는 무조건 사람이 관여함. 
mean(data, na.rm=T)
#class(data)

[1] 9.4375

[1] 7.55

In [41]:
# 1) NA 를 0으로 대체해서 평균값 구하기.

mean(ifelse(is.na(data), 0, data))


[1] 7.55

In [58]:
# 2) NA를 평균으로 대체해서 평균값 구하기.
# 아무 의미 없는 데이터.
# 평균값으로 들어갔더니 평균값이 나옴.
library(stringr)
mean(ifelse(is.na(data), mean(data, na.rm=T), data))


[1] 9.4375

In [61]:
# 3) NA를 중앙값 으로 대체해서 평균값 구하기.
mean(ifelse(is.na(data), median(data, na.rm=T), data))

[1] 9.25

In [59]:
# 4) NA를 최댓값으로 대체해서 평균값 구하기.
mean(ifelse(is.na(data), max(data, na.rm=T), data))


[1] 11.35

In [60]:
# 5) NA를 최소값으로 대체해서 평균값 구하기.
mean(ifelse(is.na(data), min(data, na.rm=T), data))

[1] 7.75

> 결측치를 제외한 평균 :9.4375<br>
> 0으로 대체 : 7.55 <br>
> 평균값으로 대체 : 9.4375<br> // 데이터가 아님. 평균은 수식?
> 중앙값으로 대체 : 9.25<br> ***
> 최대값으로 대체 : 11.35<br>
> 최소값으로 대체 : 7.75